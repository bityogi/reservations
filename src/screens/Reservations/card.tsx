import React from 'react';
import {
  View,
  Text,
  StyleSheet,
  TouchableHighlight
} from 'react-native';
import PropTypes from 'prop-types';

import { formatDate } from '../../util';
import { ReservationType } from '../../util/types'

const styles = StyleSheet.create({
  container: {
    backgroundColor: '#fff',
    flex: 1,
    padding: 10,
    paddingTop: 10,
    paddingBottom: 10,
    margin: 10,
    marginTop: 5,
    marginBottom: 5,
    borderRadius: 6,
    // borderColor: '#48BBEC',
    // borderWidth: 0.5,
  },
  name: {
    fontSize: 30,
    fontWeight: '500'
  },
  hotelName: {
    fontSize: 20,
    color: '#bdbdbd'
  },
  datesContainer: {

  },
  dates: {

  }
})

interface CardProps {
  reservation: ReservationType,
  onPress: (id: string) => void
}

const ReservationCard = ({ reservation, onPress }: CardProps) => {
  const { name, hotelName, arrivalDate, departureDate } = reservation;
  return (
    <TouchableHighlight onPress={() => onPress(reservation.id)}>
      <View style={styles.container}>
        <Text style={styles.name}>{name}</Text>
        <Text style={styles.hotelName}>{hotelName}</Text>
        <View style={styles.datesContainer}>
          <Text style={styles.dates}>{`${formatDate(arrivalDate)} to ${formatDate(departureDate)}`}</Text>
        </View>
      </View>
    </TouchableHighlight>
  )
}

ReservationCard.propTypes = {
  reservation: PropTypes.shape({
    name: PropTypes.string.isRequired,
    hotelName: PropTypes.string.isRequired,
    arrivalDate: PropTypes.instanceOf(Date),
    departureDate: PropTypes.instanceOf(Date),
  })
}

export default ReservationCard;