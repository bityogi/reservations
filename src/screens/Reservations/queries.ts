import gql from 'graphql-tag';
import { client } from '../../util/requests';
import { ReservationType } from '../../util/types';

export const getReservations = async () => {
  const query = gql`
    query loadReservations {
      reservations(first: 25, orderBy: createdAt_DESC) {
        id
        name
        hotelName
        arrivalDate
        departureDate
      }
    }
  `;
  const { data: { reservations }} = await client.query({ 
    query,
    fetchPolicy: 'no-cache'
  });

  return reservations.map((r : ReservationType) => ({
    ...r,
    arrivalDate: new Date(r.arrivalDate),
    departureDate: new Date(r.departureDate)
  }));
} 