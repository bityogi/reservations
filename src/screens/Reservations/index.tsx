import React, { PureComponent } from 'react';
import {
  StyleSheet,
  FlatList
} from 'react-native';
import ActionButton from 'react-native-action-button';



import ReservationCard from './card';
import { getReservations } from './queries';
import { ReservationType, NavigationProp } from '../../util/types';

const styles = StyleSheet.create({
  list: {
    paddingTop: 20,
    backgroundColor: '#F3F3F3'
  },
});

interface ReservationsState {
  reservations: ReservationType[]
}

class Reservations extends PureComponent<NavigationProp, ReservationsState> {
  
  state = {
    reservations: []
  }

  componentDidMount() {
    getReservations()
      .then(reservations => {
        this.setState({ reservations });
      });
  }

  willFocus = this.props.navigation.addListener(
    'willFocus',
    () => {
      const refresh = this.props.navigation.getParam('refresh', false)
      if (refresh) {
        getReservations()
          .then(reservations => {
            this.setState({ reservations });
          });
      }
    }
  );

  handleAddReservation = () => {
    const { navigation } = this.props;
    navigation.navigate('NewReservation');
  }

  onReservationDetailSelect = (reservationId: string) => {
    const { navigation } = this.props;
    navigation.navigate('ReservationDetail', {
      reservationId
    });
  }

  render() {
    const { reservations } = this.state;

    return(
      [
        <FlatList
          key="reservationsList"
          style={styles.list}
          data={reservations}
          renderItem={({ item }) => {
            return <ReservationCard reservation={item} onPress={this.onReservationDetailSelect}/>
          }}
          keyExtractor={(reservation : ReservationType) => reservation.id}
        />,
        <ActionButton
          key="addReservation"
          onPress={this.handleAddReservation}
          buttonColor="rgba(231, 76, 60, 1)"
        />
    
      ]
        
      
    )
  }
}

export default Reservations;