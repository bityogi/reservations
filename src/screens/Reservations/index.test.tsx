import 'react-native';
import React from 'react';
// Note: test renderer must be required after react-native.
import { shallow } from 'enzyme';
import Reservations from './index';
import console = require('console');

jest.mock("NativeModules", () => ({
  UIManager: {
    RCTView: () => ({
      directEventTypes: {}
    })
  },
  KeyboardObserver: {},
  RNGestureHandlerModule: {
    attachGestureHandler: jest.fn(),
    createGestureHandler: jest.fn(),
    dropGestureHandler: jest.fn(),
    updateGestureHandler: jest.fn(),
    State: {},
    Directions: {}
  }
}));


jest.mock('../../util/requests');
jest.mock('./queries');

const navigationProp = {
  navigate: (to: string, options?: object) => { console.log('Navigating') },
  getParam: (param: string) => { return '1' },
  addListener: (listenEvent: string, payload: object) => {}
}

describe('Reservations', () => {

  it('should render correctly', () => {
    const component = shallow(<Reservations navigation={navigationProp} />);
    console.log('Reservations component: ', Reservations)
    expect(component).toMatchSnapshot()
  })
})