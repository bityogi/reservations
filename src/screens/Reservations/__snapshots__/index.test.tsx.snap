// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Reservations should render correctly 1`] = `
Array [
  <FlatList
    data={Array []}
    disableVirtualization={false}
    horizontal={false}
    initialNumToRender={10}
    key="reservationsList"
    keyExtractor={[Function]}
    maxToRenderPerBatch={10}
    numColumns={1}
    onEndReachedThreshold={2}
    removeClippedSubviews={false}
    renderItem={[Function]}
    scrollEventThrottle={50}
    style={
      Object {
        "backgroundColor": "#F3F3F3",
        "paddingTop": 20,
      }
    }
    updateCellsBatchingPeriod={50}
    windowSize={21}
  />,
  <ActionButton
    active={false}
    activeOpacity={0.85}
    autoInactive={true}
    backdrop={false}
    backgroundTappable={false}
    bgColor="transparent"
    bgOpacity={1}
    buttonColor="rgba(231, 76, 60, 1)"
    buttonText="+"
    buttonTextStyle={Object {}}
    degrees={45}
    fixNativeFeedbackRadius={false}
    key="addReservation"
    nativeFeedbackRippleColor="rgba(255,255,255,0.75)"
    offsetX={30}
    offsetY={30}
    onPress={[Function]}
    onPressIn={[Function]}
    onPressOn={[Function]}
    outRangeScale={1}
    position="right"
    resetToken={null}
    size={56}
    spacing={20}
    useNativeFeedback={true}
    verticalOrientation="up"
  />,
]
`;
